### PR TITLE
Edges Codec

### PR DESCRIPTION
**Pull request by falkanson**
This adds the Edges codec, and refactors the common.py by putting encode_variant and decode_variant int it.

The edges codec works by generating an edges mask from one channel. In the pixels that are edges the other two Channels will be used for the data using the LSB method.

### CODE REVIEW

File: src/pydis_jam23/codecs/edges.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    alternative_trys = 0
+
+    # check if the channels is big enugh for the message
+    # try the other ones if not
+    while True:
+        edges = get_edges(image.split()[mask_color]).tobytes()
+        if count_color(edges, 0) > len(data):  # count edge pixels
+            break
+        elif alternative_trys >= 2:
+            msg = "The message is to long to be encoded into this image."
```
Comment (Artemis21) [inline src/pydis_jam23/codecs/edges.py]:
```suggestion
            msg = "The message is too long to be encoded into this image."
```

**Comment by falkanson** [comment]
So this should be working now...

do i have to add anything to make the code compatible with the gui?

and another important thing is that Error PLR2004 for magic values is not in pyproject.toml not sure if that is the best solution.

**Comment by falkanson** [comment]
also tested it works over CLI

File: src/pydis_jam23/codecs/edges.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+cli_flag = "--edges"
+cli_help = "use the edges codec"
+
+
+def encode(image: Image.Image, message: bytes) -> None:
+    """encode a message into an image using the above described method"""
+    image_data = bytearray(image.tobytes())
+    data = encode_varint(len(message)) + message
+
+    mask_color = randint(0, 2)  # noqa: S311 no need for cryto
```
Comment (Artemis21) [inline src/pydis_jam23/codecs/edges.py]:
```suggestion
    mask_color = randint(0, 2)  # noqa: S311 no need for crypto
```
Also: image is not guaranteed to have 3 channels. Instead consider doing `image.split()` here which returns a tuple of channels, and iterating over that (with a random offset ig?).

File: src/pydis_jam23/codecs/edges.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        if count_color(edges, 0) > len(data):  # count edge pixels
+            break
+        elif alternative_trys >= 2:
+            msg = "The message is to long to be encoded into this image."
+            raise CodecError(msg)
+        alternative_trys += 1
+        mask_color = (mask_color + alternative_trys) % 3
+
+    for idx in range(0, 3):  # set all LSB of the first image to 0
+        image_data[idx] = set_lsb(image_data[idx], 0)
```
Comment (Artemis21) [inline src/pydis_jam23/codecs/edges.py]:
```suggestion
    for idx in range(0, 3):  # set all LSB of the first pixel to 0
        image_data[idx] = set_lsb(image_data[idx], 0)
```

1. What if the first pixel is detected as an edge, and this gets overwritten?
2. Needs adjusting if the image has more/less than 3 channels.

File: src/pydis_jam23/codecs/edges.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
+"""Encode a message into the given image.
+using as custom LSB variant
+
+One of the color channels is randomly seclected.
+On this channel edge detection is run.
+The pixels that are detected as edges will then be used,
+by to store data in the other two channels.
```
Comment (Artemis21) [inline src/pydis_jam23/codecs/edges.py]:
```suggestion
The pixels that are detected as edges will then be used
to store data in the other two channels.
```

File: src/pydis_jam23/codecs/edges.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        else:
+            # skip the pixels
+            pixel_index += 1
+
+    return data_indices
+
+
+def get_edges(image: Image.Image) -> Image.Image:
+    """Find the Edges in an Image and mark them
+    - black pixels mark edges"""
```
Comment (Artemis21) [inline src/pydis_jam23/codecs/edges.py]:
```suggestion
    """Find the edges in an Image and mark them.

    Black pixels mark edges.
    """
```

**Review by Artemis21 — CHANGES_REQUESTED** [review]
Looks good and well thought out.

The main thing here is that it assumes an image will have exactly three channels, but this is not guaranteed. As I understand it this codec will not work on a one channel image, so it could throw a CodecError in that case. It should work well for a four channel image though I think?

I've also suggested a couple fixes for typos in comments and docstrings, and docstring style improvements. There are more improvements that could be made here, but I'm not sure how much it matters.

File: src/pydis_jam23/codecs/edges.py (hunk 3 above)
Comment (falkanson) [reply src/pydis_jam23/codecs/edges.py]: by drawing a rectangle around the image i ensure that the border will never be detected as an edge, also i start data insertion at pixel 1 not 0

**Comment by falkanson** [comment]
going to work on the 3 channel thing this evening - or just error when there are less then ore more then 3 channels ^^

**Review by Artemis21 — APPROVED** [review]
We got there! Good work.

**Pull request merged**